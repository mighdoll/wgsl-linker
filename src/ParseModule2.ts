import {
  AbstractElem,
  ExportElem,
  FnElem,
  ImportElem,
  ImportingItem,
} from "./AbstractElems.js";
import { parseMiniWgsl } from "./ParseWgslD.js";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule2 {
  template?: string;
  exports: TextExport2[];
  fns: FnElem[];
  imports: ImportElem[];
  name: string;
  src: string;
}

export interface TextExport2 {
  name: string;
  ref: FnElem; // TODO | StructElem (| global var?)
  exp: ExportElem;
  args: string[];
  importing: ImportingItem[];
}

let unnamedModuleDex = 0;

export function parseModule2(
  src: string,
  defaultModuleName?: string
): TextModule2 {
  const parsed = parseMiniWgsl(src);
  const exports = findExports(src, parsed);
  const fns = parsed.filter((e) => e.kind === "fn") as FnElem[];
  const imports = parsed.filter((e) => e.kind === "import") as ImportElem[];
  const moduleName = undefined; // TODO parse #module

  const name = moduleName ?? defaultModuleName ?? `module${unnamedModuleDex++}`;
  return { name, exports, fns, imports, src };
}

// TODO consider how to export fields inside a struct.. currently indicated #export foo
// should we still do text capture for fields? seems better to export (and merge?) the struct itself..

function findExports(src: string, parsed: AbstractElem[]): TextExport2[] {
  const exports: TextExport2[] = [];
  parsed.forEach((elem, i) => {
    if (elem.kind === "export") {
      const next = parsed[i + 1];
      const args = elem.args ?? [];
      const { importing = [] } = elem;
      if (next?.kind === "fn") {
        const { name } = next;
        exports.push({ exp: elem, args, ref: next, name, importing });
      } else {
        console.warn(`#export what at pos: ${elem.start}? (#export not followed by fn)`);
      }
    }
  });
  return exports;
}
