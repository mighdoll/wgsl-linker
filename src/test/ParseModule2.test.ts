import { expect, test } from "vitest";
import { parseModule2 } from "../ParseModule2.js";

test("simple fn export", () => {
  const src = `
    // #export
    fn one() -> i32 {
      return 1;
    }
  `;
  const module = parseModule2(src, "my.module");
  console.log(module.exports);
  expect(module.exports.length).toBe(1);
  expect(module).toMatchSnapshot();
});

test("simple fn import", () => {
  const src = `
    // #import foo
    fn bar() { foo(); }
  `;
  const module = parseModule2(src, "my.module");
  console.log(module);
  expect(module.imports.length).toBe(1);
  expect(module.imports[0].name).toBe("foo");
  expect(module).toMatchSnapshot();
});

// test("read simple struct export", () => {
//   const exportPrefix = `// #export`;
//   const src = `
//     struct Elem {
//       sum: f32;
//     }
//   `;
//   const module = parseModule(exportPrefix + "\n" + src);
//   expect(module.exports.length).toBe(1);
//   const firstExport = module.exports[0];
//   expect(firstExport.name).toBe("Elem");
//   expect(firstExport.params).deep.equals([]);
//   expect(firstExport.src).toBe(src);
// });

// test("read #module", () => {
//   const myModule = `
//     // #module myModule
//     // #export
//     fn foo() {}
//   `;
//   const textModule = parseModule(myModule);
//   expect(textModule.name).toBe("myModule");
// });

// test("parse #export log", () => {
//   const myModule = `
//     #export log(myVar)

//     _log(myVar)
//   `;
//   const textModule = parseModule(myModule);
//   expect(textModule.exports[0].name).toBe("log");
// });

// test("parse #export log with #endInsert", () => {
//   const myModule = `
//     #export log(myVar)
//     _log(myVar);
//     #endInsert

//     fn _log(myVar: i32) {}
//   `;
//   const textModule = parseModule(myModule);
//   const firstExport = textModule.exports[0];
//   expect(firstExport.name).toBe("log");
//   expect(firstExport.src.trim()).toBe("_log(myVar);");
//   expect(firstExport.rootSrc!.trim()).toBe("fn _log(myVar: i32) {}");
// });
