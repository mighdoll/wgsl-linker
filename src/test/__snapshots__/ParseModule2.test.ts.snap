// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simple fn export 1`] = `
{
  "exports": [
    {
      "end": 16,
      "kind": "export",
      "ref": {
        "calls": [],
        "end": 59,
        "kind": "fn",
        "name": "one",
        "start": 20,
        "typeRefs": [
          {
            "end": 35,
            "kind": "typeRef",
            "name": "i32",
            "start": 32,
          },
        ],
      },
      "start": 8,
    },
  ],
  "fns": [
    {
      "calls": [],
      "end": 59,
      "kind": "fn",
      "name": "one",
      "start": 20,
      "typeRefs": [
        {
          "end": 35,
          "kind": "typeRef",
          "name": "i32",
          "start": 32,
        },
      ],
    },
  ],
  "imports": [],
  "name": "my.module",
  "src": "
    // #export
    fn one() -> i32 {
      return 1;
    }
  ",
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;

exports[`simple fn import 1`] = `
{
  "exports": [],
  "fns": [
    {
      "calls": [
        {
          "end": 38,
          "kind": "call",
          "name": "foo",
          "start": 35,
        },
      ],
      "end": 43,
      "kind": "fn",
      "name": "bar",
      "start": 24,
      "typeRefs": [],
    },
  ],
  "imports": [
    {
      "end": 19,
      "kind": "import",
      "name": "foo",
      "start": 8,
    },
  ],
  "name": "my.module",
  "src": "
    // #import foo
    fn bar() { foo(); }
  ",
  "structs": [],
  "template": undefined,
  "vars": [],
}
`;
